// victory-chart/lib/components/victory-scatter/helper-methods.js
Object.defineProperty(exports,"__esModule",{value:true});var _defaults2=require("lodash/defaults");var _defaults3=_interopRequireDefault(_defaults2);var _omit2=require("lodash/omit");var _omit3=_interopRequireDefault(_omit2);var _pick2=require("lodash/pick");var _pick3=_interopRequireDefault(_pick2);var _values2=require("lodash/values");var _values3=_interopRequireDefault(_values2);
var _victoryCore=require("victory-core");
var _scale=require("../../helpers/scale");var _scale2=_interopRequireDefault(_scale);
var _domain=require("../../helpers/domain");var _domain2=_interopRequireDefault(_domain);
var _data=require("../../helpers/data");var _data2=_interopRequireDefault(_data);function _interopRequireDefault(obj){return obj&&obj.__esModule?obj:{default:obj};}function _toConsumableArray(arr){if(Array.isArray(arr)){for(var i=0,arr2=Array(arr.length);i<arr.length;i++){arr2[i]=arr[i];}return arr2;}else{return Array.from(arr);}}exports.default=

{
getBaseProps:function getBaseProps(props,fallbackProps){
var modifiedProps=_victoryCore.Helpers.modifyProps(props,fallbackProps);
var calculatedValues=this.getCalculatedValues(modifiedProps,fallbackProps);var
data=calculatedValues.data;var style=calculatedValues.style;var scale=calculatedValues.scale;
var childProps={parent:{
style:style.parent,scale:scale,data:data,height:modifiedProps.height,width:modifiedProps.width}};

for(var index=0,len=data.length;index<len;index++){
var datum=data[index];
var eventKey=datum.eventKey;
var x=scale.x(datum.x);
var y=scale.y(datum.y);
var dataProps={
x:x,y:y,datum:datum,index:index,scale:scale,
size:this.getSize(datum,modifiedProps,calculatedValues),
symbol:this.getSymbol(datum,modifiedProps),
style:this.getDataStyles(datum,style.data)};


var text=this.getLabelText(modifiedProps,datum,index);
var labelStyle=this.getLabelStyle(style.labels,dataProps);
var labelProps={
style:labelStyle,
x:x,
y:y-labelStyle.padding,
text:text,
index:index,
scale:scale,
datum:dataProps.datum,
textAnchor:labelStyle.textAnchor,
verticalAnchor:labelStyle.verticalAnchor||"end",
angle:labelStyle.angle};

childProps[eventKey]={
data:dataProps,
labels:labelProps};

}
return childProps;
},

getCalculatedValues:function getCalculatedValues(props,fallbackProps){
var defaultStyles=props.theme&&props.theme.scatter?props.theme.scatter:
fallbackProps.style;
var style=_victoryCore.Helpers.getStyles(props.style,defaultStyles,"auto","100%");
var data=_victoryCore.Events.addEventKeys(props,_data2.default.getData(props));
var range={
x:_victoryCore.Helpers.getRange(props,"x"),
y:_victoryCore.Helpers.getRange(props,"y")};

var domain={
x:_domain2.default.getDomain(props,"x"),
y:_domain2.default.getDomain(props,"y")};

var scale={
x:_scale2.default.getBaseScale(props,"x").domain(domain.x).range(range.x),
y:_scale2.default.getBaseScale(props,"y").domain(domain.y).range(range.y)};

var z=props.bubbleProperty||"z";
return{data:data,scale:scale,style:style,z:z};
},

getDataStyles:function getDataStyles(datum,style){
var stylesFromData=(0,_omit3.default)(datum,[
"x","y","z","size","symbol","name","label"]);

var baseDataStyle=(0,_defaults3.default)({},stylesFromData,style);
return _victoryCore.Helpers.evaluateStyle(baseDataStyle,datum);
},

getLabelText:function getLabelText(props,datum,index){
var propsLabel=Array.isArray(props.labels)?
props.labels[index]:_victoryCore.Helpers.evaluateProp(props.labels,datum);
return datum.label||propsLabel;
},

getLabelStyle:function getLabelStyle(labelStyle,dataProps){var
datum=dataProps.datum;var size=dataProps.size;var style=dataProps.style;
var matchedStyle=(0,_pick3.default)(style,["opacity","fill"]);
var padding=labelStyle.padding||size*0.25;
var baseLabelStyle=(0,_defaults3.default)({},labelStyle,matchedStyle,{padding:padding});
return _victoryCore.Helpers.evaluateStyle(baseLabelStyle,datum);
},

getSymbol:function getSymbol(data,props){
if(props.bubbleProperty){
return"circle";
}
var symbol=data.symbol||props.symbol;
return _victoryCore.Helpers.evaluateProp(symbol,data);
},

getBubbleSize:function getBubbleSize(datum,props,calculatedValues){var
data=calculatedValues.data;var z=calculatedValues.z;
var getMaxRadius=function getMaxRadius(){
var minPadding=Math.min.apply(Math,_toConsumableArray((0,_values3.default)(_victoryCore.Helpers.getPadding(props))));
return Math.max(minPadding,5);
};
var zData=data.map(function(point){return point.z;});
var zMin=Math.min.apply(Math,_toConsumableArray(zData));
var zMax=Math.max.apply(Math,_toConsumableArray(zData));
var maxRadius=props.maxBubbleSize||getMaxRadius();
var maxArea=Math.PI*Math.pow(maxRadius,2);
var area=(datum[z]-zMin)/(zMax-zMin)*maxArea;
var radius=Math.sqrt(area/Math.PI);
return Math.max(radius,1);
},

getSize:function getSize(data,props,calculatedValues){
var size=void 0;
if(data.size){
size=typeof data.size==="function"?data.size:Math.max(data.size,1);
}else if(typeof props.size==="function"){
size=props.size;
}else if(data[calculatedValues.z]){
size=this.getBubbleSize(data,props,calculatedValues);
}else{
size=Math.max(props.size,1);
}
return _victoryCore.Helpers.evaluateProp(size,data);
}};
