// victory-core/lib/victory-util/helpers.js
Object.defineProperty(exports,"__esModule",{value:true});var _property2=require("lodash/property");var _property3=_interopRequireDefault(_property2);var _partial2=require("lodash/partial");var _partial3=_interopRequireDefault(_partial2);var _merge2=require("lodash/merge");var _merge3=_interopRequireDefault(_merge2);var _isFunction2=require("lodash/isFunction");var _isFunction3=_interopRequireDefault(_isFunction2);var _defaults2=require("lodash/defaults");var _defaults3=_interopRequireDefault(_defaults2);function _interopRequireDefault(obj){return obj&&obj.__esModule?obj:{default:obj};}function _defineProperty(obj,key,value){if(key in obj){Object.defineProperty(obj,key,{value:value,enumerable:true,configurable:true,writable:true});}else{obj[key]=value;}return obj;}exports.default=

{
getPadding:function getPadding(props){
var padding=typeof props.padding==="number"?props.padding:0;
var paddingObj=typeof props.padding==="object"?props.padding:{};
return{
top:paddingObj.top||padding,
bottom:paddingObj.bottom||padding,
left:paddingObj.left||padding,
right:paddingObj.right||padding};

},

getStyles:function getStyles(style,defaultStyles,height,width){// eslint-disable-line max-params
if(!style){
return(0,_defaults3.default)({parent:{height:height,width:width}},defaultStyles);
}var

data=style.data;var labels=style.labels;var parent=style.parent;
return{
parent:(0,_defaults3.default)({height:height,width:width},parent,defaultStyles.parent),
labels:(0,_defaults3.default)({},labels,defaultStyles.labels),
data:(0,_defaults3.default)({},data,defaultStyles.data)};

},

evaluateProp:function evaluateProp(prop,data,index){
return(0,_isFunction3.default)(prop)?prop(data,index):prop;
},

evaluateStyle:function evaluateStyle(style,data,index){var _this=this;
if(!Object.keys(style).some(function(value){return(0,_isFunction3.default)(style[value]);})){
return style;
}
return Object.keys(style).reduce(function(prev,curr){
prev[curr]=_this.evaluateProp(style[curr],data,index);
return prev;
},{});
},

getRange:function getRange(props,axis){
// determine how to lay the axis and what direction positive and negative are
var isVertical=axis!=="x";
var padding=this.getPadding(props);
if(isVertical){
return[props.height-padding.bottom,padding.top];
}
return[padding.left,props.width-padding.right];
},

// for components that take single datasets
getData:function getData(props){
if(props.data){
return this.formatData(props.data,props);
}
},

formatData:function formatData(dataset,props,stringMap){
if(!dataset){
return[];
}
stringMap=stringMap||{
x:this.createStringMap(props,"x"),
y:this.createStringMap(props,"y")};

var accessor={
x:this.createAccessor(props.x),
y:this.createAccessor(props.y)};


return dataset.map(function(datum){
var x=accessor.x(datum);
var y=accessor.y(datum);
var xName=typeof x==="string"?{xName:x}:undefined;
var yName=typeof y==="string"?{yName:y}:undefined;
return(0,_defaults3.default)({
// map string data to numeric values, and add names
x:typeof x==="string"?stringMap.x[x]:x,
y:typeof y==="string"?stringMap.y[y]:y},
xName,yName,datum);
});
},

createStringMap:function createStringMap(props,axis){
var stringsFromData=this.getStringsFromData(props,axis);
if(stringsFromData.length){
return stringsFromData.reduce(function(acc,string,index){
acc[string]=index+1;
return acc;
},{});
}
return null;
},

getStringsFromData:function getStringsFromData(props,axis){
if(!props.data){
return[];
}
var key=typeof props[axis]==="undefined"?axis:props[axis];
var accessor=this.createAccessor(key);
var dataStrings=props.data.
map(function(datum){return accessor(datum);}).
filter(function(datum){return typeof datum==="string";});
// return a unique set of strings
return dataStrings.reduce(function(prev,curr){
if(typeof curr!=="undefined"&&curr!==null&&prev.indexOf(curr)===-1){
prev.push(curr);
}
return prev;
},[]);
},

createAccessor:function createAccessor(key){
// creates a data accessor function
// given a property key, path, array index, or null for identity.
if((0,_isFunction3.default)(key)){
return key;
}else if(key===null||typeof key==="undefined"){
// null/undefined means "return the data item itself"
return function(x){return x;};
}
// otherwise, assume it is an array index, property key or path (_.property handles all three)
return(0,_property3.default)(key);
},

getPartialEvents:function getPartialEvents(events,index,childProps){
return events?
Object.keys(events).reduce(function(memo,eventName){
/* eslint max-params: 0 */
memo[eventName]=(0,_partial3.default)(
events[eventName],
_partial3.default.placeholder,// evt will still be the first argument for event handlers
childProps,// event handlers will have access to data component props, including data
index,// used in setting a unique state property
eventName// used in setting a unique state property
);
return memo;
},{}):
{};
},

modifyProps:function modifyProps(props,fallbackProps,themeProps){
var themeCheck=props.theme&&props.theme.props;
var themePropsObject=themeCheck&&!themeProps?props.theme.props:themeProps;

return themeCheck?
(0,_defaults3.default)({},props,{
clipWidth:props.width,
clipHeight:props.height},

themePropsObject,
fallbackProps.props):
(0,_defaults3.default)({},props,{
clipWidth:props.width,
clipHeight:props.height},
fallbackProps.props);
},

getEvents:function getEvents(events,namespace){var _this2=this;
var onEvent=function onEvent(evt,childProps,index,eventName){
if(_this2.props.events[namespace]&&_this2.props.events[namespace][eventName]){
_this2.setState(_defineProperty({},
index,(0,_merge3.default)(
{},
_this2.state[index],
_this2.props.events[namespace][eventName](evt,childProps,index))));


}
};

return events?
Object.keys(this.props.events[namespace]).reduce(function(memo,event){
memo[event]=onEvent;
return memo;
},{}):{};
},

getEventState:function getEventState(index,namespace){
return this.state[index]&&this.state[index][namespace];
}};
