// victory-chart/lib/helpers/wrapper.js
Object.defineProperty(exports,"__esModule",{value:true});var _uniq2=require("lodash/uniq");var _uniq3=_interopRequireDefault(_uniq2);var _partialRight2=require("lodash/partialRight");var _partialRight3=_interopRequireDefault(_partialRight2);var _isFunction2=require("lodash/isFunction");var _isFunction3=_interopRequireDefault(_isFunction2);var _flatten2=require("lodash/flatten");var _flatten3=_interopRequireDefault(_flatten2);var _defaults2=require("lodash/defaults");var _defaults3=_interopRequireDefault(_defaults2);
var _react=require("react");var _react2=_interopRequireDefault(_react);
var _axis=require("./axis");var _axis2=_interopRequireDefault(_axis);
var _data=require("./data");var _data2=_interopRequireDefault(_data);
var _domain=require("./domain");var _domain2=_interopRequireDefault(_domain);
var _victoryCore=require("victory-core");function _interopRequireDefault(obj){return obj&&obj.__esModule?obj:{default:obj};}function _toConsumableArray(arr){if(Array.isArray(arr)){for(var i=0,arr2=Array(arr.length);i<arr.length;i++){arr2[i]=arr[i];}return arr2;}else{return Array.from(arr);}}exports.default=


{
getData:function getData(props,childComponents){
childComponents=childComponents||_react2.default.Children.toArray(props.children);
return this.getDataFromChildren(childComponents);
},

getDomain:function getDomain(props,axis,childComponents){
var propsDomain=_domain2.default.getDomainFromProps(props,axis);
if(propsDomain){
return propsDomain;
}
childComponents=childComponents||_react2.default.Children.toArray(props.children);
return _domain2.default.cleanDomain(this.getDomainFromChildren(props,axis,childComponents),
props,
axis);
},

setAnimationState:function setAnimationState(nextProps){
if(!this.props.animate){
return;
}
if(this.props.animate.parentState){
var nodesWillExit=this.props.animate.parentState.nodesWillExit;
var oldProps=nodesWillExit?this.props:null;
this.setState((0,_defaults3.default)({oldProps:oldProps},this.props.animate.parentState));
}else{
var oldChildren=_react2.default.Children.toArray(this.props.children);
var nextChildren=_react2.default.Children.toArray(nextProps.children);var _Transitions$getIniti=







_victoryCore.Transitions.getInitialTransitionState(oldChildren,nextChildren);var _nodesWillExit=_Transitions$getIniti.nodesWillExit;var nodesWillEnter=_Transitions$getIniti.nodesWillEnter;var childrenTransitions=_Transitions$getIniti.childrenTransitions;var nodesShouldEnter=_Transitions$getIniti.nodesShouldEnter;var nodesDoneClipPathEnter=_Transitions$getIniti.nodesDoneClipPathEnter;var nodesDoneClipPathExit=_Transitions$getIniti.nodesDoneClipPathExit;

this.setState({
nodesWillExit:_nodesWillExit,
nodesWillEnter:nodesWillEnter,
childrenTransitions:childrenTransitions,
nodesShouldEnter:nodesShouldEnter,
nodesDoneClipPathEnter:nodesDoneClipPathEnter,
nodesDoneClipPathExit:nodesDoneClipPathExit,
oldProps:_nodesWillExit?this.props:null});

}
},

getAnimationProps:function getAnimationProps(props,child,index){var _this=this;
if(!props.animate){
return child.props.animate;
}
var getFilteredState=function getFilteredState(){
var childrenTransitions=_this.state&&_this.state.childrenTransitions;
childrenTransitions=_victoryCore.Collection.isArrayOfArrays(childrenTransitions)?
childrenTransitions[index]:childrenTransitions;
return(0,_defaults3.default)({childrenTransitions:childrenTransitions},_this.state);
};

var getTransitions=props.animate&&props.animate.getTransitions;
var state=getFilteredState();
var parentState=props.animate&&props.animate.parentState||state;
if(!getTransitions){
var getTransitionProps=_victoryCore.Transitions.getTransitionPropsFactory(
props,
state,
function(newState){return _this.setState(newState);});

getTransitions=(0,_partialRight3.default)(getTransitionProps,index);
}
return(0,_defaults3.default)({getTransitions:getTransitions,parentState:parentState},props.animate,child.props.animate);
},

getDomainFromChildren:function getDomainFromChildren(props,axis,childComponents){
childComponents=childComponents||_react2.default.Children.toArray(props.children);
var horizontalChildren=childComponents.some(function(child){return child.props.horizontal;});
var horizontal=props&&props.horizontal||horizontalChildren.length>0;
var currentAxis=_axis2.default.getCurrentAxis(axis,horizontal);
var getChildDomains=function getChildDomains(children){
return children.reduce(function(memo,child){
if(child.type&&(0,_isFunction3.default)(child.type.getDomain)){
var childDomain=child.props&&child.type.getDomain(child.props,currentAxis);
return childDomain?memo.concat(childDomain):memo;
}else if(child.props&&child.props.children){
return memo.concat(getChildDomains(_react2.default.Children.toArray(child.props.children)));
}
return memo;
},[]);
};

var childDomains=getChildDomains(childComponents);
var min=_victoryCore.Collection.getMinValue(childDomains);
var max=_victoryCore.Collection.getMaxValue(childDomains);
return childDomains.length===0?
[0,1]:[min,max];
},

getDataFromChildren:function getDataFromChildren(props,childComponents){
var getData=function getData(childProps){
var data=_data2.default.getData(childProps);
return Array.isArray(data)&&data.length>0?data:undefined;
};

var getChildData=function getChildData(children){
return children.map(function(child){
if(child.type&&(0,_isFunction3.default)(child.type.getData)){
var childData=child.props&&child.type.getData(child.props);
return childData;
}else if(child.props&&child.props.children){
return(0,_flatten3.default)(getChildData(_react2.default.Children.toArray(child.props.children)));
}
return getData(child.props);
});
};
childComponents=childComponents||_react2.default.Children.toArray(props.children);
return getChildData(childComponents);
},

getStackedDomain:function getStackedDomain(props,axis){
var propsDomain=_domain2.default.getDomainFromProps(props,axis);
if(propsDomain){
return propsDomain;
}var
horizontal=props.horizontal;
var ensureZero=function ensureZero(domain){
var isDependent=axis==="y"&&!horizontal||axis==="x"&&horizontal;
return isDependent?[Math.min.apply(Math,_toConsumableArray(domain).concat([0])),Math.max.apply(Math,_toConsumableArray(domain).concat([0]))]:domain;
};
var datasets=this.getDataFromChildren(props);
return ensureZero(_domain2.default.getDomainFromGroupedData(props,axis,datasets));
},

getColor:function getColor(calculatedProps,child,index){
// check for styles first
var style=calculatedProps.style;var
colorScale=calculatedProps.colorScale;
if(style&&style.data&&style.data.fill){
return style.data.fill;
}
colorScale=child.props&&child.props.colorScale?child.props.colorScale:colorScale;
var colors=Array.isArray(colorScale)?
colorScale:_victoryCore.Style.getColorScale(colorScale);
return colors[index%colors.length];
},

getChildStyle:function getChildStyle(child,index,calculatedProps){var
style=calculatedProps.style;
var role=child.type&&child.type.role;
var defaultFill=role==="group-wrapper"||role==="stack-wrapper"?
undefined:this.getColor(calculatedProps,child,index);
var childStyle=child.props.style||{};
var dataStyle=(0,_defaults3.default)({},childStyle.data,style.data,{fill:defaultFill});
var labelsStyle=(0,_defaults3.default)({},childStyle.labels,style.labels);
return{
parent:style.parent,
data:dataStyle,
labels:labelsStyle};

},

getStringsFromCategories:function getStringsFromCategories(childComponents,axis){
var stringsFromCategories=function stringsFromCategories(children){
return children.reduce(function(memo,child){
if(child.props&&child.props.categories){
return memo.concat(_data2.default.getStringsFromCategories(child.props,axis));
}else if(child.props&&child.props.children){
return memo.concat(stringsFromCategories(
_react2.default.Children.toArray(child.props.children)));

}
return memo;
},[]);
};

return stringsFromCategories(childComponents);
},

getStringsFromData:function getStringsFromData(childComponents,axis){
var stringsFromData=function stringsFromData(children){
return children.reduce(function(memo,child){
if(child.props&&child.props.data){
return memo.concat(_victoryCore.Helpers.getStringsFromData(child.props,axis));
}else if(child.type&&(0,_isFunction3.default)(child.type.getData)){var _ret=function(){
var data=(0,_flatten3.default)(child.type.getData(child.props));
var attr=axis==="x"?"xName":"yName";
return{v:memo.concat(data.reduce(function(prev,datum){
return datum[attr]?prev.concat(datum[attr]):prev;
},[]))};}();if(typeof _ret==="object")return _ret.v;
}else if(child.props&&child.props.children){
return memo.concat(stringsFromData(_react2.default.Children.toArray(child.props.children)));
}
return memo;
},[]);
};

return stringsFromData(childComponents);
},

getStringsFromChildren:function getStringsFromChildren(props,axis,childComponents){
childComponents=childComponents||_react2.default.Children.toArray(props.children);
var axisComponent=_axis2.default.getAxisComponent(childComponents,axis);
var axisStrings=axisComponent?_data2.default.getStringsFromAxes(axisComponent.props,axis):[];
var categoryStrings=this.getStringsFromCategories(childComponents,axis);
var dataStrings=this.getStringsFromData(childComponents,axis);
return(0,_uniq3.default)((0,_flatten3.default)([].concat(_toConsumableArray(categoryStrings),_toConsumableArray(dataStrings),_toConsumableArray(axisStrings))));
},

getCategories:function getCategories(props,axis){
var categories=_data2.default.getCategories(props,axis)||
this.getStringsFromChildren(props,axis);
return categories.length>0?categories:undefined;
},

getY0:function getY0(datum,index,calculatedProps){var
datasets=calculatedProps.datasets;
var y=datum.y;
var previousDataSets=datasets.slice(0,index);
var previousPoints=previousDataSets.reduce(function(prev,dataset){
return prev.concat(dataset.
filter(function(previousDatum){return datum.x instanceof Date?
previousDatum.x.getTime()===datum.x.getTime():
previousDatum.x===datum.x;}).
map(function(previousDatum){return previousDatum.y||0;}));

},[]);
return previousPoints.reduce(function(memo,value){
var sameSign=y<0&&value<0||y>=0&&value>=0;
return sameSign?memo+value:memo;
},0);
}};
