// victory-core/lib/victory-transition/victory-transition.js
Object.defineProperty(exports,"__esModule",{value:true});var _filter2=require("lodash/filter");var _filter3=_interopRequireDefault(_filter2);var _pick2=require("lodash/pick");var _pick3=_interopRequireDefault(_pick2);var _isFunction2=require("lodash/isFunction");var _isFunction3=_interopRequireDefault(_isFunction2);var _defaults2=require("lodash/defaults");var _defaults3=_interopRequireDefault(_defaults2);var _extends=Object.assign||function(target){for(var i=1;i<arguments.length;i++){var source=arguments[i];for(var key in source){if(Object.prototype.hasOwnProperty.call(source,key)){target[key]=source[key];}}}return target;};var _createClass=function(){function defineProperties(target,props){for(var i=0;i<props.length;i++){var descriptor=props[i];descriptor.enumerable=descriptor.enumerable||false;descriptor.configurable=true;if("value"in descriptor)descriptor.writable=true;Object.defineProperty(target,descriptor.key,descriptor);}}return function(Constructor,protoProps,staticProps){if(protoProps)defineProperties(Constructor.prototype,protoProps);if(staticProps)defineProperties(Constructor,staticProps);return Constructor;};}();var _react=require("react");var _react2=_interopRequireDefault(_react);
var _victoryAnimation=require("../victory-animation/victory-animation");var _victoryAnimation2=_interopRequireDefault(_victoryAnimation);
var _index=require("../victory-util/index");function _interopRequireDefault(obj){return obj&&obj.__esModule?obj:{default:obj};}function _toConsumableArray(arr){if(Array.isArray(arr)){for(var i=0,arr2=Array(arr.length);i<arr.length;i++){arr2[i]=arr[i];}return arr2;}else{return Array.from(arr);}}function _classCallCheck(instance,Constructor){if(!(instance instanceof Constructor)){throw new TypeError("Cannot call a class as a function");}}function _possibleConstructorReturn(self,call){if(!self){throw new ReferenceError("this hasn't been initialised - super() hasn't been called");}return call&&(typeof call==="object"||typeof call==="function")?call:self;}function _inherits(subClass,superClass){if(typeof superClass!=="function"&&superClass!==null){throw new TypeError("Super expression must either be null or a function, not "+typeof superClass);}subClass.prototype=Object.create(superClass&&superClass.prototype,{constructor:{value:subClass,enumerable:false,writable:true,configurable:true}});if(superClass)Object.setPrototypeOf?Object.setPrototypeOf(subClass,superClass):subClass.__proto__=superClass;}var


VictoryTransition=function(_React$Component){_inherits(VictoryTransition,_React$Component);function VictoryTransition(){_classCallCheck(this,VictoryTransition);return _possibleConstructorReturn(this,Object.getPrototypeOf(VictoryTransition).apply(this,arguments));}_createClass(VictoryTransition,[{key:"componentWillReceiveProps",value:function componentWillReceiveProps(



















nextProps){
this.setState(this.getTransitionState(this.props,nextProps));
}},{key:"getTransitionState",value:function getTransitionState(

props,nextProps){var
animate=props.animate;
if(!animate){
return{};
}else if(animate.parentState){
var oldProps=animate.parentState.nodesWillExit?props:null;
return{oldProps:oldProps};
}else{
var oldChildren=_react2.default.Children.toArray(props.children);
var nextChildren=_react2.default.Children.toArray(nextProps.children);var _Transitions$getIniti=







_index.Transitions.getInitialTransitionState(oldChildren,nextChildren);var nodesWillExit=_Transitions$getIniti.nodesWillExit;var nodesWillEnter=_Transitions$getIniti.nodesWillEnter;var childrenTransitions=_Transitions$getIniti.childrenTransitions;var nodesShouldEnter=_Transitions$getIniti.nodesShouldEnter;var nodesDoneClipPathEnter=_Transitions$getIniti.nodesDoneClipPathEnter;var nodesDoneClipPathExit=_Transitions$getIniti.nodesDoneClipPathExit;
return{
nodesWillExit:nodesWillExit,
nodesWillEnter:nodesWillEnter,
childrenTransitions:childrenTransitions,
nodesShouldEnter:nodesShouldEnter,
nodesDoneClipPathEnter:nodesDoneClipPathEnter,
nodesDoneClipPathExit:nodesDoneClipPathExit,
oldProps:nodesWillExit?props:null};

}
}},{key:"getDomainFromChildren",value:function getDomainFromChildren(

props,axis){
var getChildDomains=function getChildDomains(children){
return children.reduce(function(memo,child){
if(child.type&&(0,_isFunction3.default)(child.type.getDomain)){
var childDomain=child.props&&child.type.getDomain(child.props,axis);
return childDomain?memo.concat(childDomain):memo;
}else if(child.props&&child.props.children){
return memo.concat(getChildDomains(_react2.default.Children.toArray(child.props.children)));
}
return memo;
},[]);
};

var childComponents=_react2.default.Children.toArray(props.children);
if(props.domain&&(Array.isArray(props.domain)||props.domain[axis])){
return Array.isArray(props.domain)?props.domain:props.domain[axis];
}else{
var childDomains=getChildDomains(childComponents);
return childDomains.length===0?
[0,1]:[Math.min.apply(Math,_toConsumableArray(childDomains)),Math.max.apply(Math,_toConsumableArray(childDomains))];
}
}},{key:"render",value:function render()

{var _this2=this;
var props=this.state&&this.state.nodesWillExit?
this.state.oldProps:this.props;
var getTransitionProps=this.props.animate&&this.props.animate.getTransitions?
this.props.animate.getTransitions:
_index.Transitions.getTransitionPropsFactory(
props,
this.state,
function(newState){return _this2.setState(newState);});

var child=_react2.default.Children.toArray(props.children)[0];
var transitionProps=getTransitionProps(child);
var domain={
x:this.getDomainFromChildren(props,"x"),
y:this.getDomainFromChildren(props,"y")};

var combinedProps=(0,_defaults3.default)(
{domain:domain},transitionProps,child.props);

var animationWhitelist=props.animationWhitelist;

if(this.state&&this.state.nodesDoneClipPathExit&&this.state.nodesWillExit||
transitionProps.animate&&
transitionProps.animate.parentState&&
transitionProps.animate.parentState.nodesDoneClipPathExit&&
transitionProps.animate.parentState.nodesWillExit){
animationWhitelist=(0,_filter3.default)(props.animationWhitelist,function(list){
return list!=="clipWidth";
});
}

var propsToAnimate=animationWhitelist?
(0,_pick3.default)(combinedProps,animationWhitelist):combinedProps;
return(
_react2.default.createElement(_victoryAnimation2.default,_extends({},combinedProps.animate,{data:propsToAnimate}),
function(newProps){
var component=_react2.default.cloneElement(
child,(0,_defaults3.default)({animate:null},newProps,combinedProps));
return component;
}));


}}]);return VictoryTransition;}(_react2.default.Component);VictoryTransition.displayName="VictoryTransition";VictoryTransition.propTypes={/**
     * The animate prop specifies an animation config for the transition.
     * This prop should be given as an object.
     */animate:_react2.default.PropTypes.object,/**
     * VictoryTransition animates a single child component
     */children:_react2.default.PropTypes.node,/**
     * This prop specifies which of the child's props are safe to interpolate.
     * This props should be given as an array.
     */animationWhitelist:_react2.default.PropTypes.array};exports.default=VictoryTransition;
