// victory-chart/lib/components/victory-errorbar/helper-methods.js
Object.defineProperty(exports,"__esModule",{value:true});var _pick2=require("lodash/pick");var _pick3=_interopRequireDefault(_pick2);var _flatten2=require("lodash/flatten");var _flatten3=_interopRequireDefault(_flatten2);var _isArray2=require("lodash/isArray");var _isArray3=_interopRequireDefault(_isArray2);var _defaults2=require("lodash/defaults");var _defaults3=_interopRequireDefault(_defaults2);var _omit2=require("lodash/omit");var _omit3=_interopRequireDefault(_omit2);var _assign2=require("lodash/assign");var _assign3=_interopRequireDefault(_assign2);var _extends=Object.assign||function(target){for(var i=1;i<arguments.length;i++){var source=arguments[i];for(var key in source){if(Object.prototype.hasOwnProperty.call(source,key)){target[key]=source[key];}}}return target;};
var _victoryCore=require("victory-core");
var _scale=require("../../helpers/scale");var _scale2=_interopRequireDefault(_scale);
var _axis=require("../../helpers/axis");var _axis2=_interopRequireDefault(_axis);
var _domain=require("../../helpers/domain");var _domain2=_interopRequireDefault(_domain);
var _data=require("../../helpers/data");var _data2=_interopRequireDefault(_data);function _interopRequireDefault(obj){return obj&&obj.__esModule?obj:{default:obj};}function _toConsumableArray(arr){if(Array.isArray(arr)){for(var i=0,arr2=Array(arr.length);i<arr.length;i++){arr2[i]=arr[i];}return arr2;}else{return Array.from(arr);}}exports.default=

{
getBaseProps:function getBaseProps(props,fallbackProps){
var modifiedProps=_victoryCore.Helpers.modifyProps(props,fallbackProps);
var calculatedValues=this.getCalculatedValues(modifiedProps,fallbackProps);var
data=calculatedValues.data;var style=calculatedValues.style;var scale=calculatedValues.scale;var
groupComponent=modifiedProps.groupComponent;var height=modifiedProps.height;var width=modifiedProps.width;var borderWidth=modifiedProps.borderWidth;
var childProps={parent:{style:style.parent,scale:scale,data:data,height:height,width:width}};
for(var index=0,len=data.length;index<len;index++){
var datum=data[index];
var eventKey=datum.eventKey||index;
var x=scale.x(datum.x);
var y=scale.y(datum.y);

var dataProps={
x:x,y:y,scale:scale,datum:datum,index:index,groupComponent:groupComponent,borderWidth:borderWidth,
style:this.getDataStyles(datum,style.data),
errorX:this.getErrors(datum,scale,"x"),
errorY:this.getErrors(datum,scale,"y")};


var labelStyle=this.getLabelStyle(style.labels,dataProps);
var labelProps={
style:labelStyle,
x:x-labelStyle.padding,
y:y-labelStyle.padding,
text:this.getLabelText(modifiedProps,datum,index),
index:index,
scale:scale,
datum:dataProps.datum,
textAnchor:labelStyle.textAnchor,
verticalAnchor:labelStyle.verticalAnchor||"end",
angle:labelStyle.angle};


childProps[eventKey]={
data:dataProps,
labels:labelProps};

}
return childProps;
},

getErrorData:function getErrorData(props){
if(props.data){
if(props.data.length<1){
_victoryCore.Log.warn("This is an empty dataset.");
return[];
}

return this.formatErrorData(props.data,props);
}else{
var generatedData=(props.errorX||props.errorY)&&this.generateData(props);
return this.formatErrorData(generatedData,props);
}
},

getErrors:function getErrors(datum,scale,axis){
/**
     * check if it is asymmetric error or symmetric error, asymmetric error should be an array
     * and the first value is the positive error, the second is the negative error
     * @param  {Boolean} isArray(errorX)
     * @return {String or Array}
     */

var errorNames={x:"errorX",y:"errorY"};
var errors=datum[errorNames[axis]];
if(errors===0){
return false;
}

return(0,_isArray3.default)(errors)?
[errors[0]===0?false:scale[axis](errors[0]+datum[axis]),
errors[1]===0?false:scale[axis](datum[axis]-errors[1])]:
[scale[axis](errors+datum[axis]),scale[axis](datum[axis]-errors)];
},

formatErrorData:function formatErrorData(dataset,props){
if(!dataset){
return[];
}
var accessor={
x:_victoryCore.Helpers.createAccessor(props.x),
y:_victoryCore.Helpers.createAccessor(props.y),
errorX:_victoryCore.Helpers.createAccessor(props.errorX),
errorY:_victoryCore.Helpers.createAccessor(props.errorY)};


var replaceNegatives=function replaceNegatives(errors){
// check if the value is negative, if it is set to 0
var replaceNeg=function replaceNeg(val){return!val||val<0?0:val;};
return(0,_isArray3.default)(errors)?errors.map(function(err){return replaceNeg(err);}):replaceNeg(errors);
};

return dataset.map(function(datum){
var x=accessor.x(datum);
var y=accessor.y(datum);
var errorX=replaceNegatives(accessor.errorX(datum));
var errorY=replaceNegatives(accessor.errorY(datum));

return(0,_assign3.default)({},datum,{x:x,y:y,errorX:errorX,errorY:errorY});
});
},

getDomain:function getDomain(props,axis){
var propsDomain=_domain2.default.getDomainFromProps(props,axis);
if(propsDomain){
return _domain2.default.padDomain(propsDomain,props,axis);
}
var categoryDomain=_domain2.default.getDomainFromCategories(props,axis);
if(categoryDomain){
return _domain2.default.padDomain(categoryDomain,props,axis);
}
var dataset=this.getErrorData(props);

if(dataset.length<1){
return _scale2.default.getBaseScale(props,axis).domain();
}

var domain=this.getDomainFromData(props,axis,dataset);
return _domain2.default.cleanDomain(_domain2.default.padDomain(domain,props,axis),props);
},

getDomainFromData:function getDomainFromData(props,axis,dataset){
var currentAxis=_axis2.default.getCurrentAxis(axis,props.horizontal);
var error=void 0;
if(currentAxis==="x"){
error="errorX";
}else if(currentAxis==="y"){
error="errorY";
}
var axisData=(0,_flatten3.default)(dataset).map(function(datum){return datum[currentAxis];});
var errorData=(0,_flatten3.default)((0,_flatten3.default)(dataset).map(function(datum){
var errorMax=void 0;
var errorMin=void 0;
if((0,_isArray3.default)(datum[error])){
errorMax=datum[error][0]+datum[currentAxis];
errorMin=datum[currentAxis]-datum[error][1];
}else{
errorMax=datum[error]+datum[currentAxis];
errorMin=datum[currentAxis]-datum[error];
}
return[errorMax,errorMin];
}));

var allData=axisData.concat(errorData);
var min=Math.min.apply(Math,_toConsumableArray(allData));
var max=Math.max.apply(Math,_toConsumableArray(allData));
// TODO: is this the correct behavior, or should we just error. How do we
// handle charts with just one data point?
if(min===max){
var adjustedMax=max===0?1:max;
return[0,adjustedMax];
}
return[min,max];
},

getCalculatedValues:function getCalculatedValues(props,fallbackProps){
var defaultStyles=props.theme&&props.theme.errorbar?props.theme.errorbar:
fallbackProps.style;
var style=_victoryCore.Helpers.getStyles(props.style,defaultStyles,"auto","100%");
var assignData=_extends(_data2.default.getData(props),this.getErrorData(props));
var data=_victoryCore.Events.addEventKeys(props,assignData);
var range={
x:_victoryCore.Helpers.getRange(props,"x"),
y:_victoryCore.Helpers.getRange(props,"y")};

var domain={
x:this.getDomain(props,"x"),
y:this.getDomain(props,"y")};

var scale={
x:_scale2.default.getBaseScale(props,"x").domain(domain.x).range(range.x),
y:_scale2.default.getBaseScale(props,"y").domain(domain.y).range(range.y)};


return{data:data,scale:scale,style:style};
},

getDataStyles:function getDataStyles(datum,style){
var stylesFromData=(0,_omit3.default)(datum,[
"x","y","name"]);

var baseDataStyle=(0,_defaults3.default)({},stylesFromData,style);
return _victoryCore.Helpers.evaluateStyle(baseDataStyle,datum);
},

getLabelText:function getLabelText(props,datum,index){
var propsLabel=Array.isArray(props.labels)?
props.labels[index]:_victoryCore.Helpers.evaluateProp(props.labels,datum);
return datum.label||propsLabel;
},

getLabelStyle:function getLabelStyle(labelStyle,dataProps){var
datum=dataProps.datum;var size=dataProps.size;var style=dataProps.style;
var matchedStyle=(0,_pick3.default)(style,["opacity","fill"]);
var padding=labelStyle.padding||size*0.25;
var baseLabelStyle=(0,_defaults3.default)({},labelStyle,matchedStyle,{padding:padding});
return _victoryCore.Helpers.evaluateStyle(baseLabelStyle,datum);
}};
