// victory-chart/lib/components/victory-bar/bar.js
Object.defineProperty(exports,"__esModule",{value:true});var _extends=Object.assign||function(target){for(var i=1;i<arguments.length;i++){var source=arguments[i];for(var key in source){if(Object.prototype.hasOwnProperty.call(source,key)){target[key]=source[key];}}}return target;};var _createClass=function(){function defineProperties(target,props){for(var i=0;i<props.length;i++){var descriptor=props[i];descriptor.enumerable=descriptor.enumerable||false;descriptor.configurable=true;if("value"in descriptor)descriptor.writable=true;Object.defineProperty(target,descriptor.key,descriptor);}}return function(Constructor,protoProps,staticProps){if(protoProps)defineProperties(Constructor.prototype,protoProps);if(staticProps)defineProperties(Constructor,staticProps);return Constructor;};}();var _react=require("react");var _react2=_interopRequireDefault(_react);
var _helperMethods=require("./helper-methods");var _helperMethods2=_interopRequireDefault(_helperMethods);function _interopRequireDefault(obj){return obj&&obj.__esModule?obj:{default:obj};}function _classCallCheck(instance,Constructor){if(!(instance instanceof Constructor)){throw new TypeError("Cannot call a class as a function");}}function _possibleConstructorReturn(self,call){if(!self){throw new ReferenceError("this hasn't been initialised - super() hasn't been called");}return call&&(typeof call==="object"||typeof call==="function")?call:self;}function _inherits(subClass,superClass){if(typeof superClass!=="function"&&superClass!==null){throw new TypeError("Super expression must either be null or a function, not "+typeof superClass);}subClass.prototype=Object.create(superClass&&superClass.prototype,{constructor:{value:subClass,enumerable:false,writable:true,configurable:true}});if(superClass)Object.setPrototypeOf?Object.setPrototypeOf(subClass,superClass):subClass.__proto__=superClass;}var

Bar=function(_React$Component){_inherits(Bar,_React$Component);function Bar(){_classCallCheck(this,Bar);return _possibleConstructorReturn(this,Object.getPrototypeOf(Bar).apply(this,arguments));}_createClass(Bar,[{key:"getVerticalBarPath",value:function getVerticalBarPath(





















props,width){var
x=props.x;var y0=props.y0;var y=props.y;
var size=width/2;
return"M "+(x-size)+", "+y0+"\n      L "+(
x-size)+", "+y+"\n      L "+(
x+size)+", "+y+"\n      L "+(
x+size)+", "+y0+"\n      L "+(
x-size)+", "+y0;
}},{key:"getHorizontalBarPath",value:function getHorizontalBarPath(

props,width){var
x=props.x;var y0=props.y0;var y=props.y;
var size=width/2;
return"M "+y0+", "+(x-size)+"\n      L "+
y0+", "+(x+size)+"\n      L "+
y+", "+(x+size)+"\n      L "+
y+", "+(x-size)+"\n      L "+
y0+", "+(x-size);
}},{key:"getBarPath",value:function getBarPath(

props,width){
return this.props.horizontal?
this.getHorizontalBarPath(props,width):this.getVerticalBarPath(props,width);
}},{key:"renderBar",value:function renderBar(

path,style,events){var _props=
this.props;var role=_props.role;var clipId=_props.clipId;
return(
_react2.default.createElement("path",_extends({
d:path,
style:style,
role:role,
shapeRendering:"optimizeSpeed"},
events,{
clipPath:"url(#"+clipId+")"})));


}},{key:"render",value:function render()

{
// TODO better bar width calculation
var _props2=this.props;var events=_props2.events;var style=_props2.style;
var barWidth=_helperMethods2.default.getBarWidth(this.props);
var path=typeof this.props.x==="number"?
this.getBarPath(this.props,barWidth):undefined;
return this.renderBar(path,style,events);
}}]);return Bar;}(_react2.default.Component);Bar.propTypes={clipId:_react.PropTypes.number,datum:_react.PropTypes.object,events:_react.PropTypes.object,horizontal:_react.PropTypes.bool,index:_react.PropTypes.number,role:_react.PropTypes.string,scale:_react.PropTypes.object,style:_react.PropTypes.object,x:_react2.default.PropTypes.number,y:_react2.default.PropTypes.number,y0:_react2.default.PropTypes.number,width:_react.PropTypes.number,padding:_react.PropTypes.oneOfType([_react.PropTypes.number,_react.PropTypes.object]),data:_react.PropTypes.array};exports.default=Bar;
