// victory-chart/lib/components/victory-bar/helper-methods.js
Object.defineProperty(exports,"__esModule",{value:true});var _omit2=require("lodash/omit");var _omit3=_interopRequireDefault(_omit2);var _defaults2=require("lodash/defaults");var _defaults3=_interopRequireDefault(_defaults2);var _assign2=require("lodash/assign");var _assign3=_interopRequireDefault(_assign2);
var _victoryCore=require("victory-core");
var _data=require("../../helpers/data");var _data2=_interopRequireDefault(_data);
var _domain=require("../../helpers/domain");var _domain2=_interopRequireDefault(_domain);
var _scale=require("../../helpers/scale");var _scale2=_interopRequireDefault(_scale);function _interopRequireDefault(obj){return obj&&obj.__esModule?obj:{default:obj};}exports.default=

{

getScale:function getScale(props){var
horizontal=props.horizontal;
var range={
x:_victoryCore.Helpers.getRange(props,"x"),
y:_victoryCore.Helpers.getRange(props,"y")};

var domain={
x:_domain2.default.getDomainWithZero(props,"x"),
y:_domain2.default.getDomainWithZero(props,"y")};

var xScale=_scale2.default.getBaseScale(props,"x").domain(domain.x).range(range.x);
var yScale=_scale2.default.getBaseScale(props,"y").domain(domain.y).range(range.y);
return{
x:horizontal?yScale:xScale,
y:horizontal?xScale:yScale};

},

getBarWidth:function getBarWidth(props){var
style=props.style;var width=props.width;var data=props.data;
var padding=props.padding.left||props.padding;
var barWidth=style&&style.width||data.length===0?
8:0.3*(width-2*padding)/data.length;

return barWidth;
},

getBarPosition:function getBarPosition(props,datum,scale){
var currentAxis=props.horizontal?"x":"y";
var defaultMin=_scale2.default.getScaleType(props,currentAxis)==="log"?
1/Number.MAX_SAFE_INTEGER:0;
var yOffset=datum.yOffset||0;
var xOffset=datum.xOffset||0;
var y0=yOffset||defaultMin;
var y=datum.y+yOffset;
var x=datum.x+xOffset;
var formatValue=function formatValue(value,axis){
return datum[axis]instanceof Date?new Date(value):value;
};
return{
x:scale.x(formatValue(x,"x")),
y0:scale.y(formatValue(y0,"y")),
y:scale.y(formatValue(y,"y"))};

},

getBarStyle:function getBarStyle(datum,baseStyle){
var styleData=(0,_omit3.default)(datum,[
"xName","yName","x","y","label"]);

return _victoryCore.Helpers.evaluateStyle((0,_defaults3.default)({},styleData,baseStyle),datum);
},

getLabelStyle:function getLabelStyle(style,datum){
var labelStyle=(0,_defaults3.default)({},{
angle:datum.angle,
textAnchor:datum.textAnchor,
verticalAnchor:datum.verticalAnchor},
style);
return _victoryCore.Helpers.evaluateStyle(labelStyle,datum);
},

getLabel:function getLabel(props,datum,index){
var propsLabel=Array.isArray(props.labels)?
props.labels[index]:_victoryCore.Helpers.evaluateProp(props.labels,datum);
return datum.label||propsLabel;
},

getLabelAnchors:function getLabelAnchors(datum,horizontal){
var sign=datum.y>=0?1:-1;
if(!horizontal){
return{
vertical:sign>=0?"end":"start",
text:"middle"};

}else{
return{
vertical:"middle",
text:sign>=0?"start":"end"};

}
},

getlabelPadding:function getlabelPadding(style,datum,horizontal){
var defaultPadding=style.padding||0;
var sign=datum.y<0?-1:1;
return{
x:horizontal?sign*defaultPadding:0,
y:horizontal?0:sign*defaultPadding};

},

getCalculatedValues:function getCalculatedValues(props,fallbackProps){
var defaultStyles=props.theme&&props.theme.scatter?props.theme.scatter:
fallbackProps.style;
var style=_victoryCore.Helpers.getStyles(props.style,defaultStyles,"auto","100%");
var data=_victoryCore.Events.addEventKeys(props,_data2.default.getData(props));
var scale=this.getScale(props);
return{style:style,data:data,scale:scale};
},

getBaseProps:function getBaseProps(props,fallbackProps){
var modifiedProps=_victoryCore.Helpers.modifyProps(props,fallbackProps);var _getCalculatedValues=
this.getCalculatedValues(modifiedProps,fallbackProps);var style=_getCalculatedValues.style;var data=_getCalculatedValues.data;var scale=_getCalculatedValues.scale;var
horizontal=modifiedProps.horizontal;var width=modifiedProps.width;var height=modifiedProps.height;var padding=modifiedProps.padding;
var childProps={parent:{scale:scale,width:width,height:height,data:data,style:style.parent}};
for(var index=0,len=data.length;index<len;index++){
var datum=data[index];
var eventKey=datum.eventKey||index;
var position=this.getBarPosition(modifiedProps,datum,scale);
var dataProps=(0,_assign3.default)(
{
style:this.getBarStyle(datum,style.data),
index:index,
datum:datum,
scale:scale,
horizontal:horizontal,
padding:padding,
width:width,
data:data},

position);


var labelStyle=this.getLabelStyle(style.labels,datum);
var labelPadding=this.getlabelPadding(labelStyle,datum,horizontal);
var anchors=this.getLabelAnchors(datum,horizontal);
var labelProps={
style:labelStyle,
x:horizontal?position.y+labelPadding.x:position.x+labelPadding.x,
y:horizontal?position.x+labelPadding.y:position.y-labelPadding.y,
y0:position.y0,
text:this.getLabel(modifiedProps,datum,index),
index:index,
scale:scale,
datum:dataProps.datum,
textAnchor:labelStyle.textAnchor||anchors.text,
verticalAnchor:labelStyle.verticalAnchor||anchors.vertical,
angle:labelStyle.angle};


childProps[eventKey]={
data:dataProps,
labels:labelProps};

}
return childProps;
}};
